deploy/buildtools/buildconfigfile.py

```diff
@@ -0,0 +1,161 @@
+
+        self.args = args
+
+        global_build_config_file = None
+        with open(args.buildconfigfile, "r") as yaml_file:
+            global_build_config_file = yaml.safe_load(yaml_file)
+
+        # aws specific options
+        self.agfistoshare = global_build_config_file['agfis-to-share']
+        self.acctids_to_sharewith = global_build_config_file['share-with-accounts'].values()
+
+        # this is a list of actual builds to run
+        builds_to_run_list = global_build_config_file['builds']
+
+        build_recipes_config_file = None
+        with open(args.buildrecipesconfigfile, "r") as yaml_file:
+            build_recipes_config_file = yaml.safe_load(yaml_file)
+
+        build_farm_hosts_config_file = None
+        with open(args.buildfarmconfigfile, "r") as yaml_file:
+            build_farm_hosts_config_file = yaml.safe_load(yaml_file)
+
+        build_recipes = dict()
+        for section_name, section_dict in build_recipes_config_file.items():
+            if section_name in builds_to_run_list:
+                build_recipes[section_name] = BuildConfig(
+                    section_name,
+                    section_dict,
+                    build_farm_hosts_config_file,
+                    self,
+                    launch_time)
+
+        self.hwdb = RuntimeHWDB(args.hwdbconfigfile)
+
+        self.builds_list = list(build_recipes.values())
+        self.build_ip_set = set()
+
+    def setup(self) -> None:
+        """Setup based on the types of build farm hosts."""
+        for build in self.builds_list:
+            auto_create_bucket(build.s3_bucketname)
+
+        # check to see email notifications can be subscribed
+        get_snsname_arn()
+
+    def request_build_farm_hosts(self) -> None:
+        """Launch an instance for the builds. Exits the program if an IP address is reused."""
+
+        def categorize(seq: List[BuildConfig]) -> Dict[str, List[BuildConfig]]:
+            """Bucket build configs by their build farm host dispatcher name.
+
+            Args:
+                seq: List of build configs to categorize.
+
+            Returns:
+                Dict of lists where the key is the build farm host name and the list holds
+                the build configs associated with it.
+            """
+            class_names = list(map(lambda x: x.build_farm_host_dispatcher.__class__.__name__, seq))
+            class_builds_zipped = zip(class_names, seq)
+
+            d = defaultdict(list)
+            for c, b in class_builds_zipped:
+                d[c].append(b)
+
+            return d
+
+        cat = categorize(self.builds_list)
+
+        for bhd_class, build_farm_hosts in cat.items():
+            # batch launching build farm hosts of similar types
+            getattr(import_module("buildtools.buildfarmhostdispatcher"),
+                bhd_class).request_build_farm_hosts(list(map(lambda x: x.build_farm_host_dispatcher, build_farm_hosts)))
```

Comment:
I made a bunch of comments on this code and then realized that this is where the review workflow is nice because I wanted to rewrite them all.  I'm still thinking about this code and need to come back here with concrete suggestions but it seems like we need to use the class looked up from `build_farm_hosts` to make the `request_build_farm_hosts` call instead of having to do the `getattr()` lookup.  This will enable us to glom other BuildHostDispatchers into the PYTHONPATH and dynamically import them as mentioned elsewhere in response to one of @davidbiancolin's comments.

Reply:
One thing to keep in mind is that you don't have to call the staticmethod from the class, you can call it from an instance.  see https://stackoverflow.com/questions/136097/difference-between-staticmethod-and-classmethod

I'm beginning to think that you want the return type of `categorize` to be `List[List[BuildConfig]]` such that each element in the returned list is a sublist with all of the configs having the same type of BuildHostDispatcher